trio/_core/tests/test_instrumentation.py

```diff
@@ -0,0 +1,266 @@
+        {
+            ("before", tasks["t1"]),
+            ("schedule", tasks["t1"]),
+            ("after", tasks["t1"]),
+            ("before", tasks["t2"]),
+            ("schedule", tasks["t2"]),
+            ("after", tasks["t2"]),
+        },
+        {
+            ("before", tasks["t1"]),
+            ("after", tasks["t1"]),
+            ("before", tasks["t2"]),
+            ("after", tasks["t2"]),
+        },
+        ("after_run",),
+    ]
+    print(list(r.filter_tasks(tasks.values())))
+    check_sequence_matches(list(r.filter_tasks(tasks.values())), expected)
+
+
+def test_null_instrument():
+    # undefined instrument methods are skipped
+    class NullInstrument:
+        def something_unrelated(self):
+            pass  # pragma: no cover
+
+    async def main():
+        await _core.checkpoint()
+
+    _core.run(main, instruments=[NullInstrument()])
+
+
+def test_instrument_before_after_run():
+    record = []
+
+    class BeforeAfterRun:
+        def before_run(self):
+            record.append("before_run")
+
+        def after_run(self):
+            record.append("after_run")
+
+    async def main():
+        pass
+
+    _core.run(main, instruments=[BeforeAfterRun()])
+    assert record == ["before_run", "after_run"]
+
+
+def test_instrument_task_spawn_exit():
+    record = []
+
+    class SpawnExitRecorder:
+        def task_spawned(self, task):
+            record.append(("spawned", task))
+
+        def task_exited(self, task):
+            record.append(("exited", task))
+
+    async def main():
+        return _core.current_task()
+
+    main_task = _core.run(main, instruments=[SpawnExitRecorder()])
+    assert ("spawned", main_task) in record
+    assert ("exited", main_task) in record
+
+
+# This test also tests having a crash before the initial task is even spawned,
+# which is very difficult to handle.
+def test_instruments_crash(caplog):
+    record = []
+
+    class BrokenInstrument:
+        def task_scheduled(self, task):
+            record.append("scheduled")
+            raise ValueError("oops")
+
+        def close(self):
+            # Shouldn't be called -- tests that the instrument disabling logic
+            # works right.
+            record.append("closed")  # pragma: no cover
+
+    async def main():
+        record.append("main ran")
+        return _core.current_task()
+
+    r = TaskRecorder()
+    main_task = _core.run(main, instruments=[r, BrokenInstrument()])
+    assert record == ["scheduled", "main ran"]
+    # the TaskRecorder kept going throughout, even though the BrokenInstrument
+    # was disabled
+    assert ("after", main_task) in r.record
+    assert ("after_run",) in r.record
+    # And we got a log message
+    exc_type, exc_value, exc_traceback = caplog.records[0].exc_info
+    assert exc_type is ValueError
+    assert str(exc_value) == "oops"
+    assert "Instrument has been disabled" in caplog.records[0].message
+
+
+def test_instruments_monkeypatch():
```

Comment:
Above this point are copied unchanged from test_run. Below this point are new tests.